cpp/src/strings/repeat_strings.cu

```diff
@@ -172,14 +169,237 @@ std::unique_ptr<column> repeat_strings(strings_column_view const& input,
                              mr);
 }
 
+namespace {
+/**
+ * @brief Functor to compute string sizes and repeat the input strings, each string is repeated by a
+ * separate number of times.
+ */
+template <class Iterator>
+struct compute_size_and_repeat_separately_fn {
+  column_device_view const strings_dv;
+  column_device_view const repeat_times_dv;
+  Iterator const repeat_times_iter;
+  bool const strings_has_nulls;
+  bool const rtimes_has_nulls;
+
+  offset_type* d_offsets{nullptr};
+
+  // If d_chars == nullptr: only compute sizes of the output strings.
+  // If d_chars != nullptr: only repeat strings.
+  char* d_chars{nullptr};
+
+  __device__ int64_t operator()(size_type const idx) const noexcept
+  {
+    auto const string_is_valid = !strings_has_nulls || strings_dv.is_valid_nocheck(idx);
+    auto const rtimes_is_valid = !rtimes_has_nulls || repeat_times_dv.is_valid_nocheck(idx);
+
+    // Any null input (either string or repeat_times value) will result in a null output.
+    auto const is_valid = string_is_valid && rtimes_is_valid;
+
+    // When the input string is null, `repeat_times` and `string_size` are also set to 0.
+    // This makes sure that if `repeat_times > 0` then we will always have a valid input string,
+    // and if `repeat_times <= 0` we will never copy anything to the output.
+    auto const repeat_times = is_valid ? repeat_times_iter[idx] : size_type{0};
+    auto const string_size =
+      is_valid ? strings_dv.element<string_view>(idx).size_bytes() : size_type{0};
+
+    // The output_size is returned, and it needs to be an int64_t number to prevent overflow.
+    auto const output_size =
+      repeat_times > 0 ? static_cast<int64_t>(repeat_times) * static_cast<int64_t>(string_size)
+                       : int64_t{0};
+
+    if (!d_chars) {
+      // If overflow happen, the stored value of output string size will be incorrect due to
+      // downcasting. In such cases, the entire output string size array should be discarded.
+      d_offsets[idx] = static_cast<size_type>(output_size);
+    } else if (repeat_times > 0) {
+      auto const d_str    = strings_dv.element<string_view>(idx);
+      auto const str_size = d_str.size_bytes();
```

Comment:
Is not `str_size` and `string_size` the same value here? It would be good not to retrieve `element<string_view>(idx)` more than once if possible.